### ISSUE TITLE
Use tokens to identify users in `language`, `updateLanguage`, `theme`, and `updateTheme`

### ISSUE THREAD

**Issue by sophie-cluml** [aicers/review-web#378]
To improve the security of certain GraphQL APIs, I think there is a benefit of validating the query parameter `username` against the username extracted from the JWT authentication token. For example, I think the `language` and `updateLanguage` GraphQL API can be enhanced by verifying that the `username` provided in the query matches the one derived from the JWT token. Another possible instance would be `resetAdminPassword` GraphQL API.

(Added later) The approach described in https://github.com/aicers/review-web/issues/378#issuecomment-2588939080 is the chosen approach. The scope of this work may include https://github.com/aicers/review-web/issues/378#issuecomment-2589375084, depending on the response by Kone.

This ensures that users can act on their own data only, preventing unauthorized access. For example, this will prevent user A to view or update user B's language setting.

However, I believe we need to grant the rights to view or modify other users' data to a certain group of privileged users, such as the users who are sys-admin or security-admin. I think this can ensure flexibility for administrative purposes while maintaining robust security for regular users.

Personally, I don't think this needs an immediate action as of now, considering the on-premise environment, but I think it would be helpful to address this matter when we go cloud.

**Comment by sehkone** [comment]
Could you share your thoughts on why the token alone might not be sufficient to identify the user? Based on your reasoning, it seems that all APIs would need to verify the `username`.

**Comment by sophie-cluml** [comment]
This is the scenario where user A can update the user B's language.

1. User A logs in. As a return, user A acquires a JWT.
2. User A sends a GraphQL API request to the `updateLanguage`, using the above acquired JWT, with the request body like:

```
mutation UpdateLanguage {
    updateLanguage(username: "userB", language: { new: "en-US" })
}
```

3. As a result, user B's language is updated to en-US.

**Comment by sehkone** [comment]
I understand that the current implementation has a significant problem. It's a bug. To address this issue, I believe the API does not need to receive the username as an argument. Instead, it should extract the `username` from the token. The both APIs, `updateLanguage` and `updateTheme` should get fixed quickly.

**Comment by sehkone** [comment]
For the `resetAdminPassword`, adding a guard to restrict access to `SystemAdministrator` only should suffice.

**Comment by sehkone** [comment]
> ~For the `resetAdminPassword`, adding a guard to restrict access to `SystemAdministrator` only should suffice.~

No changes to the API are required, as it is intended for emergency cases when users forget the admin password.

**Comment by sophie-cluml** [comment]
I agree to https://github.com/aicers/review-web/issues/378#issuecomment-2588939080 regarding `updateLanguage` and `updateTheme`.

Besides, I think below GraphQL API endpoints can also have `username` parameter removed. It is because I think user A should not be able to see user B's language or theme setting.

- `language`
- `theme`

@sehkone Could you let me know what you think about these two please?

**Comment by sophie-cluml** [comment]
> No changes to the API are required, as it is intended for emergency cases when users forget the admin password.

Noted.

**Comment by sophie-cluml** [comment]
> > ~For the `resetAdminPassword`, adding a guard to restrict access to `SystemAdministrator` only should suffice.~
> 
> No changes to the API are required, as it is intended for emergency cases when users forget the admin password.

Noted.

**Comment by sehkone** [comment]
> I agree to [#378 (comment)](https://github.com/aicers/review-web/issues/378#issuecomment-2588939080) regarding `updateLanguage` and `updateTheme`.
> 
> Besides, I think below GraphQL API endpoints can also have `username` parameter removed. It is because I think user A should not be able to see user B's language or theme setting.
> 
> * `language`
> * `theme`
> 
> @sehkone Could you let me know what you think about these two please?

I missed the two APIs. They should be treated in the same way.

**Comment by BLYKIM** [comment]
Should the API accept a token parameter directly, or should it be extracted from elsewhere before use?

**Comment by sophie-cluml** [comment]
> Should the API accept a token parameter directly, or should it be extracted from elsewhere before use?

It can be extracted from the JWT.

**Comment by BLYKIM** [comment]
> It can be extracted from the JWT.

I see that the `username` parameter is being removed. Should the API now explicitly accept a `token` parameter, or should it retrieve the JWT from the request context?

**Comment by sophie-cluml** [comment]
> > It can be extracted from the JWT.
> 
> I see that the `username` parameter is being removed. Should the API now explicitly accept a `token` parameter, or should it retrieve the JWT from the request context?

We should choose the latter.
